src/Ztm.WebApi/Watchers/TransactionConfirmation/EntityRuleRepository.cs

```diff
@@ -0,0 +1,247 @@
+using Ztm.WebApi.Callbacks;
+
+namespace Ztm.WebApi.Watchers.TransactionConfirmation
+{
+    public class EntityRuleRepository : IRuleRepository
+    {
+        readonly IMainDatabaseFactory db;
+
+        public EntityRuleRepository(IMainDatabaseFactory db)
+        {
+            if (db == null)
+            {
+                throw new ArgumentNullException(nameof(db));
+            }
+
+            this.db = db;
+        }
+
+        public async Task<Rule> AddAsync(
+            uint256 transaction, int confirmations, TimeSpan waitingTime, CallbackResult successResponse, CallbackResult timeoutResponse, Callback callback, CancellationToken cancellationToken)
+        {
+            if (transaction == null)
+            {
+                throw new ArgumentNullException(nameof(transaction));
+            }
+
+            if (successResponse == null)
+            {
+                throw new ArgumentNullException(nameof(successResponse));
+            }
+
+            if (timeoutResponse == null)
+            {
+                throw new ArgumentNullException(nameof(timeoutResponse));
+            }
+
+            if (callback == null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            using (var db = this.db.CreateDbContext())
+            {
+                var watch = await db.TransactionConfirmationWatchingRules.AddAsync
+                (
+                    new TransactionConfirmationWatchingRuleModel
+                    {
+                        Id = Guid.NewGuid(),
+                        CallbackId = callback.Id,
+                        Transaction = transaction,
+                        Status = (int)RuleStatus.Pending,
+                        Confirmation = confirmations,
+                        OriginalWaitingTime = waitingTime,
+                        RemainingWaitingTime = waitingTime,
+                        SuccessData = JsonConvert.SerializeObject(successResponse),
+                        TimeoutData = JsonConvert.SerializeObject(timeoutResponse),
+                        CurrentWatchId = null,
+                    },
+                    cancellationToken
+                );
+
+                await db.SaveChangesAsync(cancellationToken);
+
+                return ToDomain(watch.Entity, callback);
+            }
+        }
+
+        public async Task ClearCurrentWatchAsync(Guid id, CancellationToken cancellationToken)
+        {
+            using (var db = this.db.CreateDbContext())
+            {
+                var rule = await db.TransactionConfirmationWatchingRules
+                    .FirstOrDefaultAsync(c => c.Id == id, cancellationToken);
+
+                if (rule == null)
+                {
+                    throw new KeyNotFoundException("The rule id is not existed.");
+                }
+
+                rule.CurrentWatchId = null;
```

Comment:
This operation should be inside repeatable read so it will fail if somebody change it value between our read and write.